homework-assignments-1/solutions/task-4.ts

```diff
@@ -0,0 +1,200 @@
+class Product {
+  public name: string;
+  public price: number;
+  public stock: number;
+
+  constructor(name: string, price: number, stock: number) {
+    this.name = name;
+    this.price = price;
+    this.stock = stock;
+  }
+
+  addStock(quantity: number) {
+    this.stock += quantity;
+  }
+
+  reduceStock(quantity: number): void | never {
+    if (quantity > this.stock) {
+      throw new Error("Not sufficient stock");
+    }
+
+    this.stock -= quantity;
+  }
+}
+
+class Customer {
+  public name: string;
+  public email: string;
+  public cart: Product[];
+  public address: string;
+
+  constructor(name: string, email: string, address: string) {
+    this.name = name;
+    this.email = email;
+    this.address = address;
+
+    this.cart = [];
+  }
+
+  addToCart(product: Product) {
+    this.cart.push(product);
+  }
+
+  removeFromCart(product: Product) {
+    this.cart.filter(({ name }) => name !== product.name);
+  }
+}
+
+abstract class Order {
+  public customer: Customer;
+  public products: Product[];
+  public totalPrice: number;
+
+  constructor(customer: Customer, products: Product[]) {
+    this.customer = customer;
+    this.products = products;
+    this.totalPrice = 0;
+  }
+
+  abstract calculateTotalPrice(): void;
+}
+
+class PhysicalOrder extends Order {
+  public shippingAddress: string;
+  public shippingCost: number;
+  public deliveryDate: Date;
+
+  constructor(
+    customer: Customer,
+    products: Product[],
+    shippingAddress: string,
+    deliveryDate: Date,
+  ) {
+    super(customer, products);
+    this.shippingCost = 50;
+    this.shippingAddress = shippingAddress;
+    this.deliveryDate = deliveryDate;
+  }
+
+  calculateTotalPrice() {
+    this.totalPrice =
+      this.shippingCost +
+      this.products.reduce((acc, product) => acc + product.price, 0);
```

Comment:
it is not good to access object members directly. use getter functions to get price